lib/active_model/serializer/fragment_cache.rb

```diff
@@ -0,0 +1,99 @@
+module ActiveModel
+  class Serializer
+    class FragmentCache
+
+      attr_reader :serializer
+
+      def initialize(adapter, serializer, options, root)
+        @root       = root
+        @options    = options
+        @adapter    = adapter
+        @serializer = serializer
+      end
+
+      def fetch
+        klass = serializer.class
+        serializers = fragment_serializer(@serializer.object.class.name, klass)
+
+        cached_serializer = serializers[:cached].constantize.new(@serializer.object)
+        non_cached_serializer = serializers[:non_cached].constantize.new(@serializer.object)
+
+        cached_hash = @adapter.class.new(cached_serializer, @options).serializable_hash
+        non_cached_hash = @adapter.class.new(non_cached_serializer, @options).serializable_hash
+
+        if @serializer.root && @adapter.class == ActiveModel::Serializer::Adapter::JsonApi
+          cached_hash_root(cached_hash, non_cached_hash)
+        else
+          non_cached_hash.merge cached_hash
+        end
+      end
+
+      private
+
+      def cached_hash_root(cached_hash, non_cached_hash)
```

Comment:
Given this method is JsonApi specific, we'd probably have to add a new method for each adapter. That's yet another argument for the polymorphic approach I mentioned above.

Also, it should be named something like `json_api_cache_root` or something like that.